swarm/src/dial_opts.rs

```diff
@@ -141,15 +147,24 @@ impl From<PeerId> for DialOpts {
 #[derive(Debug)]
 pub struct WithPeerId {
     peer_id: PeerId,
-    condition: PeerCondition,
+    condition: Option<PeerCondition>,
+    dial_conditions: DialConditions,
     role_override: Endpoint,
     dial_concurrency_factor_override: Option<NonZeroU8>,
 }
 
 impl WithPeerId {
     /// Specify a [`PeerCondition`] for the dial.
+    #[deprecated(note = "Use `only_dial_if` instead")]
     pub fn condition(mut self, condition: PeerCondition) -> Self {
-        self.condition = condition;
+        self.condition = Some(condition);
+        self
+    }
+
+    /// Specify zero, one, or more conditions under which a dial attempt is initiated.
```

Comment:
```suggestion
    /// Specify conditions under which we might not initiate a dial attempt.
```

I think the opposite is true, we will always try to initiate a dial (after all, that is the user's intent by saying `Dial`). These conditions express when we might _not_ attempt to dial.

Reply:
The documentation for the (deprecated) `PeerCondition` used the same language:

> The available conditions under which a new dialing attempt to a known peer is initiated.

Could you clarify on what you mean by the following:

> These conditions express when we might not attempt to dial.

If I specify `DialConditions::Disconnected`, I specify a condition under which to dial. We _want_ to dial under the condition that we're not connected.

Reply:
> > These conditions express when we might not attempt to dial.
> 
> If I specify `DialConditions::Disconnected`, I specify a condition under which to dial. We _want_ to dial under the condition that we're not connected.

Well, it depends on how you look at it, right :)

When the user says `ToSwarm::Dial`, they express the intent of wanting a (new) connection. Thus, it makes more sense to me to see those conditions as situations when we are not going to make one. That is also supported by the fact that we raise a `DialError` in case we didn't end up making one.

I see `DialConditions::Disconnected` as "only proceed if we are disconnected". In other words, if we look at all situations under which we make a connection as a set, then we start with the full set and every condition reduces it to a sub-set. `Disconnected` takes away all situations where we have a connection already. `NotDialing` takes away all situations where we are already dialing.

Due to how bitflags work, we need to combine them with `|` which implies that we formulate those conditions as "negatives", otherwise combining them with an OR does not join them.

It twists my brain but I think this is the correct way of looking at it.